### PR TITLE
[vcpkg bootstrap] Allow user to set VCPKG_ROOT and CMAKE_TOOLCHAIN_FILE

### PR DESCRIPTION
**Pull request by zlaski**
Allow user to set VCPKG_ROOT and CMAKE_TOOLCHAIN_FILE environment variables.  Variables can be set for the current user only, or for all users.

Also, allow user to copy VCPG.EXE to the system directory.

The script will now require elevation.

The fix only applies to Windows platforms.

### CODE REVIEW

**Comment by JackBoosY** [comment]
I strongly doubt that we need to do these things in bootstrap.
@BillyONeal Do you agree that?

**Comment by zlaski** [comment]
Just added automatic MSBuild integration (1e0b338c8b37a138e42b7e9185d20d9e1e513196); this goes together with the VCPKG_ROOT patch (2cc1b561d918d13faa1b1fba8b4b429a48275271).

**Comment by Neumann-A** [comment]
So many no merge reasons here:

a) bootstrap should not prompt.
b) bootstrap should not require admin privs
c) vcpkg spirit is not to be installed system wide
d) vcpkg shouldn't automatically integrate into msbuild ever.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
That nothing in vcpkg edits the overall system, except integrate whose purpose is to modify the system, is very much intentional design and I don't think we should change that.

We should probably make `vcpkg activate` set `CMAKE_TOOLCHAIN_FILE` like this though.

File: scripts/bootstrap.ps1 [hunk 1]
```diff
@@ -60,6 +67,36 @@ if ($LASTEXITCODE -ne 0)
...
+if ($prompt -eq "c" -or $prompt -eq "C") {
+    New-ItemProperty -Path $registryPathUser -Name "CMAKE_TOOLCHAIN_FILE" -Value "$vcpkgRootDir\scripts\buildsystems\vcpkg.cmake" -PropertyType String -Force | Out-Null
+    $Env:CMAKE_TOOLCHAIN_FILE="$vcpkgRootDir\scripts\buildsystems\vcpkg.cmake"
+}
+elseif ($prompt -eq "a" -or $prompt -eq "A") {	
+    New-ItemProperty -Path $registryPathAll -Name "CMAKE_TOOLCHAIN_FILE" -Value "$vcpkgRootDir\scripts\buildsystems\vcpkg.cmake" -PropertyType String -Force | Out-Null
+    $Env:CMAKE_TOOLCHAIN_FILE="$vcpkgRootDir\scripts\buildsystems\vcpkg.cmake"
+}
+
+$prompt = Read-Host -Prompt "Install VCPKG executable in System folder (Y=yes, N=no)"
```
Comment (BillyONeal) [inline scripts/bootstrap.ps1:93]: This is almost certainly a bad idea because the vcpkg.exe will get out of sync with the scripts tree.

**Comment by zlaski** [comment]
Added yet another small fix.

I would argue that vcpkg _should_ be installed system-wide, the same way that MSBuild/Visual Studio are.
Also, I believe that performing MSBuild integration at bootstrap time is very convenient.
And installing vcpkg in the System folder avoids the necessity of adding yet another directory to PATH, which is overburdened as it is.

But, no hard feelings if you dislike the PR. :-)

**Comment by zlaski** [comment]
But what we _really_ should have is a self-extracting installer.  It would make the out-of-the-box experience much more civilized.

**Comment by zlaski** [comment]
> a) bootstrap should not prompt.
> b) bootstrap should not require admin privs
> c) vcpkg spirit is not to be installed system wide
> d) vcpkg shouldn't automatically integrate into msbuild ever.

Point (a) does seem sensible and I could see the motivation for (b) and (c), but what is the reasoning behind (d)?

**Comment by Neumann-A** [comment]
> Point (a) does seem sensible and I could see the motivation for (b) and (c), but what is the reasoning behind (d)?

You know I have like 10 vcpkg folders for different PRs so I don't ever want any of those be integrated. 
Also MSBuild integration is kind of lacking, so no, I don't want it integrated into VS ever. CMake is the way to go ;)

File: scripts/bootstrap.ps1 [hunk 2]
```diff
@@ -60,6 +61,12 @@ if ($LASTEXITCODE -ne 0)
     throw
 }
 
+[Environment]::SetEnvironmentVariable("VCPKG_ROOT", "$vcpkgRootDir", "user")
+[Environment]::SetEnvironmentVariable("VCPKG_COMMAND", "$vcpkgRootDir\vcpkg.exe", "user")
+[Environment]::SetEnvironmentVariable("CMAKE_TOOLCHAIN_FILE", "$vcpkgRootDir\scripts\buildsystems\vcpkg.cmake", "user")
```
Comment (JackBoosY) [inline scripts/bootstrap.ps1:66]: Do all the pwsh version support this function?

**Comment by zlaski** [comment]
Will create a new PR from a branch.

**Pull request closed without merging**